Fix Flutter's reference to the `arch` binary on macOS

madsmtm opened the pull request:
## Description of changes

`darwin.adv_cmds` contains no `arch` binary, the correct binary is instead provided via. `stdenv.hostPlatform.darwinArch`, for an example of how this is done elsewhere see [moarvm](https://github.com/NixOS/nixpkgs/blob/7355353390a62108cbe4992246fdba0ef6c65ec6/pkgs/development/interpreters/rakudo/moarvm.nix#L25). This mistake was introduced in https://github.com/NixOS/nixpkgs/pull/286750#issuecomment-1935147584, CC @reckenrode.

I'm with limited internet today, so have done this through the GitHub UI and haven't tested this locally (as I'd have to clone the `nixpkgs` repo). Feel free to squash the changes with a proper Git commit message, I think I failed to follow the guidelines on that.

Pinging package maintainers @babariviere @ericdallo @FlafyDev @hacker1024.

## Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

reckenrode commented:
Sorry about causing the issue. I don’t know how that worked for me when the binary clearly isn’t there.

wegank commented:
Wait, how can this work? `stdenv.hostPlatform.darwinArch` is just a string that evaluates to either `"x86_64"` or `"arm64"`, i.e. the result of `$(/usr/bin/arch)`, not the binary itself.

madsmtm commented:
> Wait, how can this work? `stdenv.hostPlatform.darwinArch` is just a string that evaluates to either `"x86_64"` or `"arm64"`, i.e. the result of `$(/usr/bin/arch)`, not the binary itself.

Yeah, you're completely correct, I overlooked that the other places where this is used, the replaced command is `` `/usr/bin/arch` ``, with the backticks, which makes the shell it's executed in execute the command and replace with the result.

I'm not sure what the best way to fix this is then? Is there some other place where Nix provides an `arch` binary? How much does it matter that Flutter is using the system path for that binary, and would a better approach be to change this to refer to just `arch` taken from the current PATH instead?

wegank commented:
There's a trivial fix, without Rosetta support (https://github.com/apple-oss-distributions/system_cmds/commit/8579bd456573d7d205fec79af332a12e12464a60), though. 

```diff
diff --git a/arch.tproj/arch.c b/arch.tproj/arch.c
index 4152bf0..c2c24b5 100644
--- a/arch.tproj/arch.c
+++ b/arch.tproj/arch.c
@@ -75,9 +75,10 @@ static const CPUTypes knownArchs[] = {
 #if defined(__i386__) || defined(__x86_64__)
     {"i386", CPU_TYPE_I386},
     {"x86_64", CPU_TYPE_X86_64},
+#elif defined(__arm64__)
+    {"arm64", CPU_TYPE_ARM64},
 #elif defined(__arm__)
     {"arm", CPU_TYPE_ARM},
-#endif
 #else
 #error "Unsupported architecture"
 #endif
@@ -99,10 +100,12 @@ extern char **environ;
 #if defined(__i386__) || defined(__x86_64__)
 #define NATIVE_32	"i386"
 #define NATIVE_64	"x86_64"
+#elif defined(__arm64__)
+#define NATIVE_64	"arm64"
+#define NATIVE_32	NULL
 #elif defined(__arm__)
 #define NATIVE_32	"arm"
 #define NATIVE_64	NULL
-#endif
 #else
 #error "Unsupported architecture"
 #endif

```

reckenrode commented:
I was able to recover my work on updating system_cmds to 970.0.4. I’m going to open an update PR this weekend once I confirm the stdenv bootstrap builds for both Darwin platforms. I was hoping to wait until after the newer SDKs were added, but it would be better to land it now instead of adding workarounds that are fixed with an update.

reckenrode commented:
https://github.com/NixOS/nixpkgs/pull/296321 is the PR to update system_cmds. It builds `arch` on aarch64-darwin (and x86_64-darwin if it’s overridden to use a newer SDK).

madsmtm committed:
flutter: Allow impure reference to the `arch` binary on macOS

The `arch` binary is not available in `darwin.adv_cmds`, it is provided via. `darwin.system_cmds` instead, but support for this is still experimental.

madsmtm commented:
I've gotten the project cloned locally, but just running `nix-build -A darwin.system_cmds` doesn't seem to provide the `arch` binary, so this still doesn't work.

Would you mind a lot if I just changed this PR to use `/usr/bin/arch`, until this is resolved fully?

wegank commented:
> I've gotten the project cloned locally, but just running `nix-build -A darwin.system_cmds` doesn't seem to provide the `arch` binary, so this still doesn't work.

That PR was landed on staging, it'll take weeks for it to land on master.

> Would you mind a lot if I just changed this PR to use `/usr/bin/arch`, until this is resolved fully?

I'm OK with that.

madsmtm commented:
> That PR was landed on staging, it'll take weeks for it to land on master.

Ah, makes sense.

> I'm OK with that.

Cool, I've done that then.
